src/org/labkey/remoteapi/query/SaveRowsApiCommand.java

```diff
@@ -0,0 +1,288 @@
+ *  updateCmd.setAuditBehavior(SaveRowsCommand.AuditBehavior.DETAILED);
+ *  updateCmd.setAuditUserComment("Updated promoter region coordinates based on new assembly");
+ *  saveCmd.addCommand(updateCmd);
+ *
+ *  // Delete obsolete annotation
+ *  saveCmd.addCommand(new Command(CommandType.Delete, "genome", "GeneAnnotations",
+ *      List.of(Map.of("name", "Putative enhancer", "geneName", "MYC"))));
+ *
+ *  // Execute all commands in a transaction
+ *  SaveRowsApiResponse response = saveCmd.execute(conn, "GenomeProject");
+ * </code></pre>
+ */
+public class SaveRowsApiCommand extends PostCommand<SaveRowsApiResponse>
+{
+    private final List<Command> _commands = new ArrayList<>();
+    private Map<String, Object> _extraContext;
+    private Boolean _transacted;
+    private Boolean _validateOnly;
+
+    public SaveRowsApiCommand(Command... commands)
+    {
+        super("query", "saveRows.api");
+        addCommand(commands);
+    }
+
+    public Map<String, Object> getExtraContext()
+    {
+        return _extraContext;
+    }
+
+    public SaveRowsApiCommand setExtraContext(Map<String, Object> extraContext)
+    {
+        _extraContext = extraContext;
+        return this;
+    }
+
+    public SaveRowsApiCommand addCommand(Command... commands)
+    {
+        for (Command command : commands)
+        {
+            if (command != null)
+                _commands.add(command);
+        }
+
+        return this;
+    }
+
+    public List<Command> getCommands()
+    {
+        return _commands;
+    }
+
+    public Boolean isTransacted()
+    {
+        return _transacted;
+    }
+
+    public SaveRowsApiCommand setTransacted(Boolean transacted)
+    {
+        _transacted = transacted;
+        return this;
+    }
+
+    public Boolean isValidateOnly()
+    {
+        return _validateOnly;
+    }
+
+    public SaveRowsApiCommand setValidateOnly(Boolean validateOnly)
+    {
+        _validateOnly = validateOnly;
+        return this;
+    }
+
+    @Override
+    public JSONObject getJsonObject()
+    {
+        JSONObject json = new JSONObject();
+
+        List<JSONObject> commands = new ArrayList<>();
+        for (Command command : getCommands())
+            commands.add(command.getJsonObject());
+        json.put("commands", commands);
+
+        if (getExtraContext() != null && !getExtraContext().isEmpty())
+            json.put("extraContext", getExtraContext());
+
+        if (isTransacted() != null)
+            json.put("transacted", isTransacted());
+
+        if (isValidateOnly() != null)
+            json.put("validateOnly", isValidateOnly());
+
+        return json;
+    }
+
+    @Override
+    protected SaveRowsApiResponse createResponse(String text, int status, String contentType, JSONObject json)
+    {
+        return new SaveRowsApiResponse(text, status, contentType, json);
+    }
+
+    public enum CommandType
+    {
+        Insert,
+        Update,
+        Delete
+    }
+
+    public static class Command
```

Comment:
Largely a duplicate of `SaveRowsCommand`, but for the `CommandType`.  Even if we decide not to rebrand `SaveRowsCommand`, I think making that class a shell that derives from `CommandBase` that could then also be used here might be a good idea.